### ISSUE TITLE
Found some bugs and errors by PVS-Studio analyzer

### ISSUE THREAD

**Issue by ComeInRage** [OGRECave/ogre#2405]
Hello,
I have found some bugs using PVS-Studio tool. PVS-Studio is a static code analyzer for C, C++,C# and Java:
•	Here is the [link](https://pvs-studio.com/en/blog/posts/cpp/0927/) to the article about all bugs found
•	Open Source projects [can use](https://pvs-studio.com/en/blog/posts/0614/%C2%A0) PVS-Studio as free

There are some bugs below:
[V614](https://pvs-studio.com/en/docs/warnings/v614/) Uninitialized variable 'lodLevel.reductionValue' used. main.cpp 806
Code of LodLevel structure:
```
struct _OgreLodExport LodLevel
{
  // ....
  VertexReductionMethod reductionMethod;
  Real reductionValue;
  // ....
};
```
Code fragment, uses one:
```
numLod = opts.numLods;
LodLevel lodLevel;            // <=
lodLevel.distance = 0.0;
for (unsigned short iLod = 0; iLod < numLod; ++iLod)
{

  lodLevel.reductionMethod = opts.usePercent
    ? LodLevel::VRM_PROPORTIONAL
    : LodLevel::VRM_CONSTANT;

  if (opts.usePercent)
  {
    lodLevel.reductionValue += opts.lodPercent * 0.01f;    // <=
  }
  else
  {
    lodLevel.reductionValue += (Ogre::Real)opts.lodFixed;  // <=
  }

  lodLevel.distance += opts.lodDist;
  lodConfig.levels.push_back(lodLevel);
}
```
[V595](https://pvs-studio.com/en/docs/warnings/v595/) The 'params' pointer was utilized before it was verified against nullptr. Check lines: 95, 101. OgreGpuProgramManager.cpp 95
```
Resource* GpuProgramManager::createImpl(...., 
                                        const NameValuePairList* params)
{
  auto langIt = params->find("language");
  auto typeIt = params->find("type");

  if (langIt == params->end())
    langIt = params->find("syntax");

  if (!params || langIt == params->end() || typeIt == params->end()) // <=
  {
    OGRE_EXCEPT(Exception::ERR_INVALIDPARAMS,
      "You must supply 'language' or 'syntax' and 'type' parameters");
  }
}
```
[V609](https://pvs-studio.com/en/docs/warnings/v609/). Possible division or mod by zero. OgreInstanceBatchHW_VTF.cpp 392
```
static const uint16 c_maxTexWidthHW = 4096;
const size_t numBones = 
  std::max<size_t>(1, baseSubMesh->blendIndexToBoneIndexMap.size());

// ....

const size_t maxUsableWidth = c_maxTexWidthHW –
                             (c_maxTexWidthHW % (numBones * mRowLength));

// ....

size_t texHeight = numWorldMatrices * mRowLength / maxUsableWidth; // <=
```

Best regards,
Grigory Semenchev

**Comment by paroj** [comment]
thanks for your nice write-up of the code analyzer results.

I have fixed most of the raised points in 00288c6468a68f501fd96853ea40f757c5d1b1c3 and 6cdf53b0c79d6fc05dca5d39b1542602a62cdd1d

Note, that you should probably mention that static code analysis should be used in tandem with unit-tests. That seemingly trivial off-by one in `AnimationTrack::_buildKeyFrameIndexMap` turned out to be a misunderstanding of `std::lower_bound` and led to 096955d180f45ae76eb216045b864008fde745b5

**Comment by paroj** [comment]
last remaining issue fixed in 364456b5b2fea72bd4901dbcda3b8ab5a757cf26
